packages/netlify/src/index.ts

```diff
@@ -1,2 +1,220 @@
+	 * ---
+	 * <p>{Date.now()}</p>
+	 * ```
+	 */
+	cacheOnDemandPages?: boolean;
+}
+
+export default function netlifyIntegration(
+	integrationConfig?: NetlifyIntegrationConfig
+): AstroIntegration {
+	let _config: AstroConfig;
+	let outDir: URL;
+	let rootDir: URL;
+	let astroMiddlewareEntryPoint: URL | undefined = undefined;
+
+	const ssrOutputDir = () => new URL('./.netlify/functions-internal/ssr/', rootDir);
+	const middlewareOutputDir = () => new URL('.netlify/edge-functions/middleware/', rootDir);
+
+	const cleanFunctions = async () =>
+		await Promise.all([clearDirectory(middlewareOutputDir()), clearDirectory(ssrOutputDir())]);
+
+	async function writeRedirects(routes: RouteData[], dir: URL) {
+		const fallback = _config.output === 'static' ? '/.netlify/static' : '/.netlify/functions/ssr';
+		const redirects = createRedirectsFromAstroRoutes({
+			config: _config,
+			dir,
+			routeToDynamicTargetMap: new Map(
+				routes
+					.filter(isStaticRedirect) // all other routes are handled by SSR
+					.map((route) => {
+						// this is needed to support redirects to dynamic routes
+						// on static. not sure why this is needed, but it works.
+						route.distURL ??= route.redirectRoute?.distURL;
+
+						return [route, fallback];
+					})
+			),
+		});
+
+		if (!redirects.empty()) {
+			await appendFile(new URL('_redirects', outDir), '\n' + redirects.print() + '\n');
+		}
+	}
+
+	async function writeSSRFunction() {
+		await writeFile(
+			new URL('./ssr.mjs', ssrOutputDir()),
+			`
+				import createSSRHandler from './entry.mjs';
+				export default createSSRHandler(${JSON.stringify({
+					cacheOnDemandPages: Boolean(integrationConfig?.cacheOnDemandPages),
+				})});
+				export const config = { name: "Astro SSR", generator: "@astrojs/netlify@${packageVersion}", path: "/*", preferStatic: true };
+			`
+		);
+	}
+
+	async function writeMiddleware(entrypoint: URL) {
+		await mkdir(middlewareOutputDir(), { recursive: true });
+		await writeFile(
+			new URL('./entry.mjs', middlewareOutputDir()),
+			`
+						import { onRequest } from "${fileURLToPath(entrypoint)}";
+						import { createContext, trySerializeLocals } from 'astro/middleware';
+
+						export default async (request, context) => {
+							const ctx = createContext({ 
+								request,
+								params: {}
+							});
+							ctx.locals = { netlify: { context } }
+							const next = () => {
+								const { netlify, ...otherLocals } = ctx.locals;
+								request.headers.set("x-astro-locals", trySerializeLocals(otherLocals));
+								return context.next();
+							};
+						
+							return onRequest(ctx, next);
+						}
+
+						export const config = {
+							name: "Astro Middleware",
+							generator: "@astrojs/netlify@${packageVersion}",
+							path: "/*", excludedPath: ["/_astro/*", "/.netlify/images/*"]
+						};
+						`
+		);
+
+		// taking over bundling, because Netlify bundling trips over NPM modules
```

Comment:
This was the major talking point over removing edge adapters to begin with. It's not just Netlify, any build tool would trip over NPM packages. Sooner or later every user would find a package that can't automatically be transformed.

The fact that edge middleware was opt-in kept these issues few and far between. I'm noticing that this is being enabled by default, which opens us up to that maintenance burden again.

This will likely be blocking.

Reply:
Thanks for flagging this in advance. I'd really like to have this enabled by default, because from my perspective Middleware only really makes sense when it runs close to the user - so it needs to be an edge function.

I hear the maintenance argument loud and clear. There's two things that should reduce the burden from this change:

First, we've started building NPM support for Edge Functions. It's currently in Beta (see [this post](https://answers.netlify.com/t/support-for-npm-modules-in-edge-functions/101645)), and already covers a bunch of usecases. This is maintained by us, so any bugs in there are on us. If we can make use of that, errors during bundling would visibly be shown as part of Netlify's build step, not as part of the framework build. This should move a sizeable part of bug reports towards Netlify's support, instead of the Astro folks.

Secondly - there will always be issues that are opened in this repository, or in the Astro Discord, that will place a burden on the Astro Team. But we're happy to help! We've agreed to help maintain this adapter - part of that is responding to issues etc. If you tag us on issues regarding NPM modules in Edge Functions, we'll happily deal with them! Same is true for Discord - you might have seen me jumping into some questions in the #netlify channel. I still need to figure out how to set up email notifications for that, but feel free to tag me on questions there as well - i'll be happy to help.

Does that make you feel OK with enabling Edge Functions by default?

The reason I put ESBuild into here, is that there's a standing bug around namespaced imports (like `astro/middleware`). I'll take a look at fixing those, that should also allow us to remove the esbuild step here - so there's one less thing to maintain regarding NPM packages.

Reply:
My **main** concern is around **billing**. In the past, we enabled Vercel features that affected user's tiers without knowing it. We reverted those changes. Maintenance burden comes second place, from my point of view.

I want to make sure that users know what they're doing. While Netlify offers `1M /month` of Edge Functions to the users when they use the Free Tier, I don't know how to translate these values, and maybe even the user doesn't know. Build billing is measured in time, so it's straightforward for the users, although the measure of the Edge Functions is a bit different, and it might be difficult to translate/understand for rookies (I am a rookie).

Reply:
I understand your concern around billing, but in the case of Edge Functions it should be fine. Let me put that number into perspective:

1M / month means 1 million requests that run an edge function. Since Astro middleware runs on all pages, but not on static assets, that's 1 Million page views essentially. That's enough to cover one page view every 3 seconds, which should be more than enough for sites on the free tier.

If a site gets super popular, every million requests on top of that free million is 2$ - so even if your site gets one visit per second (super popular site already!), that's 4$ per month in overages.

At these numbers, I don't think anybody on the free tier would run into overages. And our billing data shows the same, it's *extremely* rare that a free tier customer uses more than 1M edge function executions.

Reply:
> I'll take a look at fixing those, that should also allow us to remove the esbuild step here - so there's one less thing to maintain regarding NPM packages.

I'll have to walk back on this one. I gave it a go in https://github.com/netlify/edge-bundler/pull/547, but this is very hard to build within the architecture we currently have for NPM support. We'll have to rethink our approach there, or find a way to piggy-back on Deno's NPM support.

Let's stick with the ESBuild step we have in there for now, and replace it once we have something.

Reply:
From my perspective, this feature is already opt in. If you don't use middleware in your project, no middleware will be deployed to Netlify. It's only when you use middleware that an Edge Function is deployed.

*If* a user writes middleware, we'll have to deploy that somewhere. That's either Functions or Edge Functions, and looking at [our pricing page](https://www.netlify.com/pricing/#core-pricing-table) you can see that Edge Functions is the cheaper option of the two. Plus it runs on static pages, which (as per Astro's documentation) is required for Middleware.

You've probably noticed i'm a bit picky on this one, because I want the adapter no more complex than necessary. The more config options, the more questions will come up and the bigger the support burden.

Could you outline what opt-in mechanism you would like to see to alleviate this? Happy to make any changes necessary for that.

Reply:
Let's remove the billing aspect and the advantages of running code on the edge. I don't doubt the statements.

There are some technical aspects

An edge function runs with a different runtime (Deno), and if the user uses code that isn't Deno-compatible (mostly some dependencies), this will be come source of errors. A user should be able to opt-out of edge middleware if this happens. 

Plus 

> not all users need access to Netlify context
> not all users require having edge middleware

That I mentioned before.

> Could you outline what opt-in mechanism you would like to see to alleviate this? 

The previous configuration was enough. It's just a flag that says:

```js
netlify({
	edgeMiddleware: false
})
```

Reply:
I can see the confusion! 😅 The middleware is **available** in both modes, but how it's used/executed is different (for reasons!) based on the mode.

The middleware in SSG is **executed** during the page generation. Once the page is generated, the middleware isn't there anymore, because we have a static page. If a user wants to have a protected static page, Astro (middleware) isn't the way. 

Instead in SSR, the middleware is executed at each request, and that's when a user might decide to have a protected page with authorization.

Reply:
Got it! This is not obvious from the docs. I see it says “ 
every time a page or endpoint is about to be *rendered*”, which technically complies with the behaviour you’re outlining - but considering how “Middleware” works in other frameworks, it would be good to explicitly call this out.

Right now, the docs show an Authentication usecase (see [here](https://docs.astro.build/en/guides/middleware/#chaining-middleware)), which you’re saying does *not* work without Edge Middleware. This adds to the confusion.

I’ll work on reintroducing the `edgeMiddleware` option later today. Should it be opt-in or opt-out?

This brings me to another thing that we might need to build: If I’m understanding you right, any Response headers added in a Middleware won’t be present on static pages, because it writes the files to disk and the headers are lost. Is there some way for adapters to hook into that and get access to these custom headers? We could emit them headers into the `_headers` file, so the CDN sets them on static assets, similar to `_redirects`.

Reply:
I've readded `edgeMiddleware` as opt-in in https://github.com/withastro/adapters/pull/84/commits/bd7dbba6c23e815f23bd73a69b43bf0504d81e00. Chose to go opt-in because the Vercel adapter also does that, and it's good to behave the same by default across different adapters.

There's a lot of nuance around the behaviour for prerendered pages here. With `edgeMiddleware: false`, locals are applied to rendering, but you can't implement access control using it - with `edgeMiddleware: true`,  you can implement access control but locals don't work. This is confusing for all SSR adapters, and it could be useful to explain more about the tradeoffs here in the Astro docs.

Reply:
Just to check in here, if you still think Astro docs are confusing here, I would suggest open a Issue directly in the docs repo or discuss this in the docs channel on Discrod. I'm afraid, it'll get lost here!